### ISSUE TITLE
[Question] What's planned for this mod?

### ISSUE THREAD

**Issue by MythrilBagels** [Sploder12/Potioncraft#1]
(TL:DR it's in the title)

Hi! I recently found this mod of yours on an average "scroll through Modrinth sorted by recently updated" and had my interest piqued. Cauldron brewing is a concept I have been interested in for a while, even coming up with my own concept. Unfortunately, I cannot code for shit. 
But before I get too excited, I wanted to hear what sort of content/features you have planned for the future, what overall direction you plan to take with this mod.

**Comment by Sploder12** [comment]
Thanks for the interest!

This mod is meant to be fairly small and feel like it's part of vanilla (The idea originally came from Bedrock's Cauldrons [with some influence from Thaumcraft 4's Crucible]).
My current plan is, in no particular order,

- Allow creating tipped arrows
- Allow ingredients to be thrown into the cauldron (currently only right-click)
- Make it possible to start brewing directly in the cauldron

If you have any ideas, I'd be happy to hear them!

**Comment by MythrilBagels** [comment]
Oh believe me, I have many ideas lmao
My own concept focused on turning the Cauldron into what it could have been if they kept the original idea.
(For context, when Cauldrons were first added, they were going to be the main way of making potions, however playtesting deemed it unintuitive, so they instead added the Brewing Stand.)

For starters, with my idea the Cauldron would likely need to be remade from scratch due to how different it would be from the vanilla one.
I imagine this new Cauldron would have a single item slot and a fluid tank (technically, not physically).
It would have a GUI like every other crafting station in the game, however completely optional. This UI would solely show information, such as what ingredients are in the slot, what fluid is inside, the progress bar, and a recipe book. This Cauldron would still be interacted with primarily in-world, right click the block with bottles, buckets, items, etc.
The recipes must be data-driven, cause screw the brewing stand and its hardcoded potion recipes.
Important note regarding the internal fluid tank, it would be amazing if it could also hold modded fluids, **_for maximum custom recipe potential!_**

I also had a very specific idea in mind for recipe book/potion functionality, if that is something you're interested in.
Basically, instead of picking up an item and unlocking all the recipes that use that item, the book would first tell the player that it can be used, but not what for. Only after actually using the item to make a potion, would it tell you what potion it makes, to encourage experimentation.

Misc Notes:
- By replacing the default Cauldron, care would need to be taken to keep its vanilla usage such as dripstone lava, rain water, powdered snow, etc.

These are my demands. Failure to comply will result in... uh... me coping, I guess
I'm free to discuss any of these ideas further, if you want to

**Comment by Sploder12** [comment]
Those are some really great ideas! 

I really like the idea of using other fluids, especially modded fluids! Doing it data-driven would help out with that and improve customizability, that's definitely on my todo list now!

I don't really like relying on GUIs but I'll consider it depending on how well I can represent special fluids outside the GUI. Thaumcraft's Crucible did a really good job without using a GUI, and I'd like to at least try to achieve the same results. 

The recipe book idea is really nice, the experimentation aspect would really add depth to this mod. Making a custom book would take a lot of time, but it'd be worth it. I could use something like Patchouli but I don't really like the look of books made with it and adding dependencies is something I'd like to avoid if possible.

It'll take me a while, but I shall comply with your demands!

**Comment by MythrilBagels** [comment]
Sick! I'm glad you're interested in my ideas, I've been putting a lot of effort into them.

Regarding the recipe book, you mentioned using Patchouli for it, while I was just thinking of integration with the vanilla recipe book, the little icon on the GUI's that you can toggle.

Speaking of GUI, I did also say that the GUI would be optional, and only serve the purpose of showing more information (such as the aforementioned recipe book), since I understand how important the whole interactive feel of the default Cauldron is. I even made a mockup of what this GUI might look like.
![image](https://github.com/Sploder12/Potioncraft/assets/52741226/dc1e4771-8575-43f0-ad45-d0938a247c33)
It's got a single slot for an item, a progress bar, an un-interactable output slot that shows what fluid is inside, a little bubble indicator to show that its actually working (since I imagine brewing potions this way should take longer than the brewing stand), and the little flame icon, I thought it might be neat if you provide heat underneath, like a magma block or campfire, to speed up the brewing process.

Either way, I'm excited for the future of this mod! Keep up the great work :D

**Comment by Ahydul** [comment]
I'm going to comment some of my suggestions too if you don't mind.

First of all, I find the fact that the cauldron gets brewed/mixed instantly unsatisfying. I believe it would be better if there was some sort of configurable delay and some effects that shows the potion is finished. Adding to this, it may be interesting exploring ideas to make the brewing process faster by using heat (campfire would make it a bit faster, lava would make it even faster) or using a stick to stir, for example.

Another thing I found is that the brewing is not currently capped. I'm not sure if that is by design but I believe there should be some sort of limit (configurable) to the amount of ingredients you can add to a potion. For example you could spoil a potion if you saturate it with too many ingredients making it a mundane potion. Although I also think that brewing with a cauldron should be a bit more powerful than when using the brewing stand, which can be automated.

I really like the different brewing recipes you've added. I can't figure out how to make an invisibility potion. I would suggest a way to configure the recipes and add the ability to create your own recipes with mod effects.

Lastly I just wanted to say that I don't really like the GUI idea neither. If it's added I would appreciate if it's optional.

You've made a really good mod!

**Comment by Sploder12** [comment]
Thank you for the input!

I agree, that's at the top of my implementation priorities! Heat is already present too so that wouldn't be much work to adjust. 

My original idea was to mathematically cap the potions, ie. you would need an absurd amount of an item to get more effect. But balancing is difficult... So the crating potential idea might be a good route to explore.

I based the default recipes on vanilla, with some tweaks to the spider eye. I felt night vision becoming blindness felt more natural than it becoming invisibility, but I totally missed that that made invisibility uncraftable! There is a way to configure the recipes and it does work with modded effects, I just haven't gotten around to properly documenting how to do it...

Thanks!

**Comment by MythrilBagels** [comment]
> Lastly I just wanted to say that I don't really like the GUI idea neither. If it's added I would appreciate if it's optional.

In what way do you not like the idea of a GUI? This question applies to Sploder as well. As I said in my initial pitch, the GUI should be optional, and would only serve to display certain bites of information otherwise inaccessible to the Player, such as progress, contained fluid, recipes, etc.

Also regarding the potency mechanic, I will be honest and say that I am not sure whether I am a fan. I would personally much rather have it just use one ingredient, with the drawback being that you can't brew stronger/longer potions, or splash/lingering potions.

**Comment by Ahydul** [comment]
> In what way do you not like the idea of a GUI? This question applies to Sploder as well. As I said in my initial pitch, the GUI should be optional, and would only serve to display certain bites of information otherwise inaccessible to the Player, such as progress, contained fluid, recipes, etc.

It's just an opinion. Simply put I find it more immersive without a GUI. The cauldron was supposed to be used like that but Mojang decided that was not intuitive enough and introduced the brewing stand. I think that was a mistake, it's a boring mechanic.

**Comment by Sploder12** [comment]
> > Lastly I just wanted to say that I don't really like the GUI idea neither. If it's added I would appreciate if it's optional.
> 
> In what way do you not like the idea of a GUI? This question applies to Sploder as well. As I said in my initial pitch, the GUI should be optional, and would only serve to display certain bites of information otherwise inaccessible to the Player, such as progress, contained fluid, recipes, etc.
> 
> Also regarding the potency mechanic, I will be honest and say that I am not sure whether I am a fan. I would personally much rather have it just use one ingredient, with the drawback being that you can't brew stronger/longer potions, or splash/lingering potions.
> 
> Edit: Also wanted to add that by having regular potions made in the cauldron, could possibly fix the issue of modifying them in a brewing stand breaking them.

My GUI thoughts are pretty much the same as Ahydul's. Although, an optional GUI would be a nice accessibility feature to have. But, it is extremely difficult to implement well, especially keeping track of progress (i.e. if all the recipes change).

I think doing the potency mechanic with a configurable potency would be the best of both worlds. You could configure it to only allow one ingredient or set it to a practically infinite amount, with some reasonable amount being the default. Furthermore, you could configure each individual recipe to use more or less potency.

In regards to only making regular potions, doing so would be a major rework and also limit the mod significantly.

**Comment by MythrilBagels** [comment]
I'm no expert when it comes to making mods, but I feel I understand the basics of programming a blockEntity, isn't variable progress quite easy to keep track of? The furnace blocks do it, something about displaying the progress bar as a percentage of the completed time, I think?

Also, you say you wanted your mod to feel like it was a part of vanilla, but by keeping the potency mechanic, and making potions crafted this way incompatible with the brewing stand, I feel goes directly against that. It makes it feel too separated (Not to mention how brokenly overpowered it is.)

This is just me dreaming at this point, but I wanted to see a cauldron that works very similarly to the brewing stand, minus the ability to modify the potions, which would still be a Brewing Stand-only mechanic, the cauldron only being used to brew the simplest form of the potions.

I still appreciate the effort you've been pouring into this mod, however. It is very admirable!
Maybe someday there will be a mod that scratches this needlessly specific itch of mine.

**Comment by Sploder12** [comment]
My bad, I meant the recipe book Progression, not the brewing progress. Brewing progress is as you say, quite easy.

The potency would aim to put a cap on the power so that it doesn't become brokenly overpowered. Custom potions being incompatible with the brewing stand is an issue with buggy vanilla code, which can be fixed using mixins.

Using the customizability of potency, that dream should be achievable.

**Comment by MythrilBagels** [comment]
Would I, in theory, be able to brew a custom output fluid? And then collect it with that custom fluids dedicated container?
That's the part that I'm _really_ after

**Comment by Sploder12** [comment]
That is certainly something I would like to support, but there are of course challenges to that

**Comment by karetArts** [comment]
just made an account what are these abstract default profile pictures 😭

cool mod. To the guy that asked for a delay on brewing in the cauldron to make it more satisfying, how about just adding a sound effect whenever you add an ingredient to add meaning to the action

also im not really sure if this is supposed to happen 
![image_2024-02-10_193828960](https://github.com/Sploder12/Potioncraft/assets/57247411/f604df06-1984-484c-b296-a7e532ad6873)
trying to turn a crafted potion into a splash removes all of its effects

therfore its impossible to make posion harming splash potions to murder creepers with 😔

**Comment by Sploder12** [comment]
> just made an account what are these abstract default profile pictures 😭
> 
> 
> 
> cool mod, its basically exactly the idea i wanted to make real for cauldrons so i appreciate you and your modding ability a lot. To the guy that asked for a delay on brewing in the cauldron to make it more satisfying, how about just adding a sound effect whenever you add an ingredient to add meaning to the action
> 
> 
> 
> also im not really sure if this is supposed to happen 
> 
> ![image_2024-02-10_193828960](https://github.com/Sploder12/Potioncraft/assets/57247411/f604df06-1984-484c-b296-a7e532ad6873)
> 
> trying to turn a crafted potion into a splash removes all of its effects
> 
> 
> 
> therfore its impossible to make posion harming splash potions to murder creepers with 😔

Thank you!

Sound effects should be present in the latest release. Unless there is some bug I missed.

The Splash Potion issue is a known bug inherent to how vanilla brewing stands work (or more of don't work). I will make a patch when I have the time.
